code/iaas/auth-logic/src/main/java/io/cattle/platform/iaas/api/auth/dao/impl/AuthDaoImpl.java

```diff
@@ -129,70 +160,221 @@ public void updateAccount(Account account, String name, String kind, String exte
+        //Database call.
+        Condition allMembers = DSL.falseCondition();
+        for (ExternalId id: externalIds){
+            allMembers = allMembers.or(PROJECT_MEMBER.EXTERNAL_ID.eq(id.getId())
+                    .and(PROJECT_MEMBER.EXTERNAL_ID_TYPE.eq(id.getType()))
+                    .and(PROJECT_MEMBER.REMOVED.isNull())
+                    .and(PROJECT_MEMBER.STATE.eq(CommonStatesConstants.ACTIVE)));
+        }
+        SelectQuery query = create().selectQuery();
+        query.addFrom(ACCOUNT);
+        query.addJoin(PROJECT_MEMBER, PROJECT_MEMBER.PROJECT_ID.equal(ACCOUNT.ID));
+        query.addConditions(allMembers);
+        query.setDistinct(true);
+        projects.addAll(query.fetchInto(ACCOUNT));
         return projects;
-        
+    }
+
+    public List<? extends ProjectMember> getProjectMembers(Long id) {
+        return create()
+                .selectFrom(PROJECT_MEMBER)
+                .where(PROJECT_MEMBER.PROJECT_ID.eq(id))
+                .and(PROJECT_MEMBER.STATE.eq(CommonStatesConstants.ACTIVE))
+                .and(PROJECT_MEMBER.REMOVED.isNull())
+                .orderBy(PROJECT_MEMBER.ID.asc()).fetch();
+    }
+
+    @Override
+    public ProjectMember getProjectMember(Long id) {
+        return create()
+                .selectFrom(PROJECT_MEMBER)
+                .where(PROJECT_MEMBER.ID.eq(id)).fetchOne();
+    }
+
+    @Override
+    public Boolean hasAccessToProject (Long projectId, Long policyAccount, Boolean isAdmin, Set<ExternalId> externalIds) {
+        return isProjectMember(projectId, policyAccount, isAdmin, externalIds);
     }
     
-    public ObjectManager getObjectManager() {
-        return objectManager;
+    public Boolean isProjectOwner (Long projectId, Long policyAccount, Boolean isAdmin, Set<ExternalId> externalIds){
+        if (projectId == null || externalIds == null) {
+            return false;
+        }
+        if (isAdmin) {
+            return true;
+        }
+        if (policyAccount.equals(projectId)){
+            return false;
+        }
+        Set<ProjectMemberRecord> projectMembers = new HashSet<>();
+        Condition allMembers = DSL.falseCondition();
+        for (ExternalId id: externalIds){
+            allMembers = allMembers.or(PROJECT_MEMBER.EXTERNAL_ID.eq(id.getId())
+                    .and(PROJECT_MEMBER.EXTERNAL_ID_TYPE.eq(id.getType()))
+                    .and(PROJECT_MEMBER.ROLE.eq(ProjectConstants.OWNER))
+                    .and(PROJECT_MEMBER.PROJECT_ID.eq(projectId))
+                    .and(PROJECT_MEMBER.STATE.eq(CommonStatesConstants.ACTIVE))
+                    .and(PROJECT_MEMBER.REMOVED.isNull()));
+        }
+        projectMembers.addAll(create().selectFrom(PROJECT_MEMBER).where(allMembers).fetch());
+        return !projectMembers.isEmpty();
+    }
+    
+    public Boolean isProjectMember (Long projectId, Long policyAccount, Boolean isAdmin,Set<ExternalId> externalIds){
```

Comment:
Same here. Could you please convert return type, and projectId, policyAccount, isAdmin to primitives?